### PR TITLE
2555 Add support for multiple valuesets and system filters to ValueSetExpander

### PR DESCRIPTION
**Pull request by ewoutkramer**
Expander now supports having more than one valueset in an include and specifying a valueset + system to limit the import of said valuesets.

Also reworked the expander to more closely follow the expansion algorithm specified in the spec.

- Using more delayed processing, so processing valuesets that are too big is actually stopped before attempting it.
- Documented all steps more clearly, referring to the spec.
- specifying both System and ValueSet will now do an intersection.
- Cyclic references between composes are detected

### CODE REVIEW

**Comment by ewoutkramer** [comment]
Note: this PR only contains the changes for STU3, for ease of review.

PR #2559 will make sure this functionality also exists for R4+, but doing that in this PR would make reviewing hard due to reorganizations in the directory structure.

TL;DR:  first review this PR, then 2559 and if both are ok, merge this one first, then 2559.

File: src/Hl7.Fhir.Shims.STU3AndUp/Specification/Terminology/ValueSetExpander.cs [hunk 1]
```diff
@@ -0,0 +1,348 @@
+﻿#nullable enable
```
Comment (ewoutkramer) [inline src/Hl7.Fhir.Shims.STU3AndUp/Specification/Terminology/ValueSetExpander.cs:1]: This is a dead file, it is not part of the project and is not compiled. No review necessary (it's a copy of the other ValueSetExpander.cs anywat).

File: src/Hl7.Fhir.STU3/Specification/Terminology/ValueSetExpander.cs [hunk 2]
```diff
@@ -72,127 +93,164 @@ private void setExpansionParameters(ValueSet vs)
...
-            if (!conceptSet.ValueSet.Any() && conceptSet.System == null)
-                throw Error.InvalidOperation($"Encountered a ConceptSet with neither a 'system' nor a 'valueset'");
+            if (conceptSet.ValueSetElement.Any())
+            {
+                // > valueSet(s) only: Codes are 'selected' for inclusion if they are in all the referenced value sets
+                // "all the referenced sets" means we need to calculate the intersection of the expanded valuesets.
+                var expanded = await T.Task.WhenAll(conceptSet.ValueSet.Select(vs => expandValueSetAndFilterOnSystem(vs))).ConfigureAwait(false);
+                var concepts = expanded.Length == 1 ? expanded.Single() : expanded.Aggregate((l, r) => l.Intersect(r, _systemAndCodeComparer));
+
+                addCapped(result, concepts, $"Import of valuesets '{string.Join(",", conceptSet.ValueSet)}' would result in an expqansion larger than the maximum expansion size.");
```
Comment (mmsmits) [inline src/Hl7.Fhir.STU3/Specification/Terminology/ValueSetExpander.cs:162]:
```suggestion
                addCapped(result, concepts, $"Import of valuesets '{string.Join(",", conceptSet.ValueSet)}' would result in an expansion larger than the maximum expansion size.");
```

**Review by mmsmits — CHANGES_REQUESTED** [review]

File: src/Hl7.Fhir.Base/Specification/Terminology/ValueSetExpanderSettings.cs [hunk 3]
```diff
@@ -57,13 +63,14 @@ public void CopyTo(ValueSetExpanderSettings other)
             other.MaxExpansionSize = MaxExpansionSize;
             other.ValueSetSource = ValueSetSource;
             other.IncludeDesignations = IncludeDesignations;
+            // other.IncludeInactive = IncludeInactive;
```
[the remark this answers is not in the document]
Comment (ewoutkramer) [reply src/Hl7.Fhir.Base/Specification/Terminology/ValueSetExpanderSettings.cs:66]: We probably have to. But it was too much yak-shaving at this point in time, so I changed my mind and let it be.

[the remark this answers is not in the document]
Comment (ewoutkramer) [reply src/Hl7.Fhir.Base/Specification/Terminology/ValueSetExpanderSettings.cs:66]: Actually, it's too hard to impelement at this time, or anytime soon, so I'll remove it.

File: src/Hl7.Fhir.STU3/Specification/Terminology/ValueSetExpander.cs [hunk 4]
```diff
@@ -72,127 +93,164 @@ private void setExpansionParameters(ValueSet vs)
...
+        // It processes mainly two parts, which each return 0..* expanded ContainsComponents:
+        // * The "System" group (system + filter + concepts).
+        // * The "ValueSet" group (valueset)
+        // The results of both of these parts are then intersected.
+        private async T.Task<List<ValueSet.ContainsComponent>> processConceptSet(ValueSet.ConceptSetComponent conceptSet, Stack<string> inclusionChain)
         {
-            if (source.Compose == null) return;
+            // vsd-1
+            if (!conceptSet.ValueSetElement.Any() && conceptSet.System == null)
+                throw Error.InvalidOperation($"Encountered a ConceptSet with neither a 'system' nor a 'valueset'");
```
[the remark this answers is not in the document]
Comment (ewoutkramer) [reply src/Hl7.Fhir.STU3/Specification/Terminology/ValueSetExpander.cs:129]: It's not an error. According to http://hl7.org/fhir/valueset.html#combinations, a valueset may have both a compose, expand, one of the two, or none ;-)

I'll document that there.

[the remark this answers is not in the document]
Comment (mmsmits) [reply src/Hl7.Fhir.STU3/Specification/Terminology/ValueSetExpander.cs:129]: I actually meant the vsd-1, we can enforce it with an error (like you do), but we can also just return without throwing an exception (act stupid). 
What would a user expect?

[the remark this answers is not in the document]
Comment (ewoutkramer) [reply src/Hl7.Fhir.STU3/Specification/Terminology/ValueSetExpander.cs:129]: This is existing behaviour, so I think we should let it in.

[the remark this answers is not in the document]
Comment (ewoutkramer) [reply src/Hl7.Fhir.STU3/Specification/Terminology/ValueSetExpander.cs:129]: (in the old code the enforcement of this rule was spread over multiple lines, but I reshuffled it so it becomes really explicit)

**Review by mmsmits — APPROVED** [review]
